statsrelay/src/backend_client/prom_client.rs

```diff
@@ -0,0 +1,522 @@
+                name: String::from("__name__"),
+                value: String::from_utf8(id.name.to_owned()).unwrap_or_default(),
+            });
+
+            let samples = samples
+                .iter()
+                .map(|owned| Sample {
+                    value: owned.value().to_owned(),
+                    timestamp: 0,
+                    sample_rate: owned.sample_rate().unwrap_or_default(),
+                })
+                .collect();
+
+            TimeSeries {
+                labels,
+                samples,
+                exemplars: vec![],
+            }
+        })
+        .collect();
+
+    let metadata: Vec<MetricMetadata> = timeseries_map
+        .iter()
+        .map(|(id, _)| MetricMetadata {
+            r#type: MetricType::from(id.mtype).into(),
+            metric_family_name: String::from_utf8(id.name.to_owned()).unwrap_or_default(),
+            help: String::from(""),
+            unit: String::from(""),
+        })
+        .collect();
+
+    WriteRequest {
+        timeseries,
+        metadata,
+    }
+}
+
+fn compress_write_request(write_request: WriteRequest) -> Result<Vec<u8>, snap::Error> {
+    let encoded = write_request.encode_to_vec();
+    // This compression shouldn't fail given our use case. An error is returned if
+    // the total number of bytes to compress exceeds 2^32 - 1, but we limit the size
+    // of the request.
+    snap::raw::Encoder::new().compress_vec(&encoded)
+}
+
+fn should_retry(res: &Result<reqwest::Response, reqwest::Error>) -> bool {
+    match res {
+        Ok(_) => false,
+        Err(e) => match e.status() {
+            Some(s) => s.is_server_error(),
+            None => false,
+        },
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::io::Cursor;
+
+    use bytes::Bytes;
+    use hyper::http::response::Builder;
+
+    use crate::{stats::Collector, statsd_proto};
+
+    use super::*;
+
+    #[derive(Clone, Default)]
+    struct MockPromWriteRequestClient {
+        responses: Arc<Mutex<Vec<Vec<u8>>>>,
+    }
+
+    #[async_trait]
+    impl PromWriteRequestClient for MockPromWriteRequestClient {
+        async fn send_write_request(
+            &self,
+            compressed_write_request: Vec<u8>,
+        ) -> Result<reqwest::Response, reqwest::Error> {
+            self.responses
+                .lock()
+                .unwrap()
+                .push(compressed_write_request);
+            let r: reqwest::Response = Builder::new().status(200).body("{}").unwrap().into();
+            r.error_for_status()
+        }
+    }
+
+    #[allow(clippy::ptr_arg)]
+    fn decode_response(compressed: &Vec<u8>) -> WriteRequest {
+        let decompressed = snap::raw::Decoder::new()
+            .decompress_vec(compressed)
+            .expect("failed snappy decompression");
+        WriteRequest::decode(Cursor::new(decompressed)).expect("failed decoding write request")
+    }
+
+    #[tokio::test]
+    async fn test_prom_client() {
```

Comment:
This feels like a good candidate for quickcheck, if you have a structural representation of statsd pdus you can use for comparison. It's fine as is, with the by-hand test, but with protocol stuff like this I always feel a little more confident I've got all the edge cases when it's automated.

Reply:
I finally took the time to learn how to use quickcheck and it's great!